src/window.cpp

```diff
@@ -895,57 +874,100 @@ static bool MayBeShown(const Window *w)
  * @param top Top edge of the rectangle that should be repainted
  * @param right Right edge of the rectangle that should be repainted
  * @param bottom Bottom edge of the rectangle that should be repainted
+ * @param gfx_dirty Whether to mark gfx dirty
  */
-static void DrawOverlappedWindow(Window *w, int left, int top, int right, int bottom)
-{
-	const Window *v;
-	FOR_ALL_WINDOWS_FROM_BACK_FROM(v, w->z_front) {
-		if (MayBeShown(v) &&
-				right > v->left &&
-				bottom > v->top &&
-				left < v->left + v->width &&
-				top < v->top + v->height) {
-			/* v and rectangle intersect with each other */
-			int x;
-
-			if (left < (x = v->left)) {
```

Comment:
If it isn't valid to use this construct in `DrawDirtyViewport`, then it isn't valid to use it here either. Let's nip this in the bud before anybody else copypastas it :smiley:

I also took the opportunity to optimize it somewhat; hopefully the compiler is smart enough to recognise that the lambda is nothing more than a nested (recursive) function and doesn't bother with all the `std::function` indirections it might otherwise perform.